### ISSUE TITLE
Matcher.groupCount behaviour inconsistent between JS and JVM

### ISSUE THREAD

**Issue by nrinaudo** [scala-js/scala-js#3280]
The following works fine under both platforms:

```scala
val matcher = Pattern.compile("\\d+").matcher("2147483647")

matcher.find()

matcher.groupCount
```

The following fails in `js` but not `jvm`:

```scala
val matcher = Pattern.compile("\\d+").matcher("2147483647")

//matcher.find()

matcher.groupCount
```

There's nothing in the `Matcher` scaladoc that seems to imply that calling `find` before `groupCount` is needed. 

It does, however, seem like an unfixable bug. I'm no javascript expert, but it does rather look that regexes can only expose the group count for a given match, not for a matcher.

**Comment by sjrd** [comment]
Indeed, we need an existing match in JS to be able to obtain the group count. We could *invent* a matching string to produce a match, but producing such a string is non-trivial. However, we could in theory do a shallow parse of the regex to count the number of non-capturing groups.

I'll leave that open and up for grabs by the community.

**Comment by neVERberleRfellerER** [comment]
Since
https://www.ecma-international.org/ecma-262/6.0/#sec-regexpbuiltinexec
points 19 nad 20 are always hit and set result size to NcapturingParens+1 and
https://www.ecma-international.org/ecma-262/6.0/#sec-notation
states that "NcapturingParens is the total number of left capturing parentheses" then it should be actually easier to just create all-matching regexp by prepending "|". I have it here
https://github.com/neVERberleRfellerER/scala-js/commit/720a36b44bc9adc774f17138aa06e8eb9beae8e9
Tests passing

**Comment by sjrd** [comment]
PR welcome 🙂

**Comment by neVERberleRfellerER** [comment]
PR to master or 0.6.x? I am not sure if fix of this is breaking change. I can imagine that there is code somewhere in dark corner of the Galaxy that uses exception from calling groupCount to check, if find has been already called or not, calling find in catch and retrying.

**Comment by sjrd** [comment]
0.6.x. It doesn't matter that it could be breaking someone's code, since we're aligning with the JVM. That always goes to 0.6.x.

**Comment by sjrd** [comment]
Fixed in e5d3d6b495d9aaf4e2b3aadbe24598071317242c.
